doc/tutorials/debug.rst

```diff
@@ -0,0 +1,323 @@
+      pr_fatal("fatal...level %d", LOG_FATAL);     //level 1
+
+If the built-in logging doesn't provide enough information, you can add
+additional logging in functions you want to debug, using the functions
+noted above.  For example, add the following code into function
+``shell_cmd_help`` in the source file
+``acrn-hypervisor/hypervisor/debug/shell.c``:
+
+.. figure:: images/debug_image22.png
+   :align: center
+
+   shell_cmd_help added information
+
+We set console log level to 5, and mem log level to 2 through the
+command::
+
+   loglevel 5 2
+
+Then we input "help" into the ACRN console, and find the log as follows.
+
+.. figure:: images/debug_image23.png
+   :align: center
+
+   console log information
+
+Then we use the command, on the ACRN console::
+
+   sos_console
+
+to switch to the SOS console the use the commands::
+
+   systemctl enable acrnlog
+   reboot
+
+to enable the mem log and reboot the SOS.
+
+Next we set the console and mem log levels with::
+
+   loglevel 5 2
+
+and run the command::
+
+   help
+
+Finally we get into the SOS and use the command::
+
+   cat /tmp/acrnlog/acrnlog_cur.0
+
+on the console, and we'll see the following log:
+
+.. figure:: images/debug_image24.png
+   :align: center
+
+   mem log information
+
+When we have enabled NPK in the hypervisor, SOS, and UOS, we can use the
+following commands to get the NPK log:
+
+1. Start NPK output in SOS::
+
+      $ echo 1 > /sys/bus/intel_th/devices/0-msc0/active
+
+2. Wait a moment, stop NPK output, and get the log from memory::
+
+      $ echo 0 > /sys/bus/intel_th/devices/0-msc0/active
+      $ cat /dev/intel_th0/msc0 > npk_trace
+
+3. Use `PVT
+   <https://ubit-artifactory.intel.com/artifactory/list/dsdjf-pvt-repo/1909/Red/InstallPVT.All.1.0.9006.600.exe>`_
+   tool to load the NPK log:
+
+   .. figure:: images/debug_image26.png
+      :align: center
+
+      load NPK log
+
+   Then you will see the NPK log as follows:
+
+   .. figure:: images/debug_image27.png
+      :align: center
+
+      NPK log
+
+ACRN Trace
+**********
+
+ACRN trace is a tool running on the Service OS (SOS) to capture trace
+data. We can use the existing trace information to analyze, and we can
+add self-defined tracing to analyze code which we care about.
+
+Using Existing trace event id to analyze trace
+==============================================
+
+As an example, we can use the existing vm_exit trace to analyze the
+reason and times of each vm_exit after we have done some operations.
+
+1. Run the following SOS console command to collect
+   trace data::
+
+      root@clearlinux_acrn:latest /home/trace/acrntrace/20190219-001529/# acrntrace -c
+
+2. Check current directory, and confirm the directory contains four
+   trace files::
+
+      root@clearlinux_acrn:latest /home/trace/acrntrace/20190219-001529/# ls
+      0 1 2 3
+
+3. Use the command to get a summary of vmexit::
+
+      root@clearlinux_acrn:latest /# ./acrnalyze.py \
```

Comment:
same here:
```
# ./acrnalyze.py -i /home/trace/acrntrace/20190219-001529/1 -o vmexit --vm_exit
```

Reply:
We can modify it as follows:
3. Use the command to get a summary of vmexit:
`# ./acrnalyze.py -i /home/trace/acrntrace/20190219-001529/1 -o vmexit --vm_exit`
Note:
``/home/trace/acrntrace/20190219-001529`` is acrntrace genrating path
acrnalyze.py script is in ``acrn-hypervisor/tools/acrntrace/scripts/``